File: 00002_add_two_numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        node1 = l1
        node2 = l2
        overflow = 0
        currNode = None
        
        while (node1 is not None or node2 is not None or overflow != 0):
            v1 = node1.val if node1 else 0
            v2 = node2.val if node2 else 0
            v = v1 + v2 + overflow
            overflow = 1 if v >= 10 else 0
            v = v - 10 if overflow == 1 else v

            if currNode is None:
                currNode = ListNode(v)
                res = currNode
            else:
                currNode.next = ListNode(v)
                currNode = currNode.next
                
            
            
            # go next
            node1 = node1 and node1.next
            node2 = node2 and node2.next
            
        return res

File: 00002_add_two_numbers/solution.go
tion for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {

    var node1 *ListNode = l1;
    var node2 *ListNode = l2;
    var overflow = 0;
    var currNode *ListNode = nil;
    var res *ListNode;


    for (node1 != nil || node2 != nil || overflow != 0) {

        var v1, v2 int;
        if node1 != nil {
            v1 = node1.Val;
        } else {
            v1 = 0;
        }
        if node2 != nil {
            v2 = node2.Val;
        } else {
            v2 = 0;
        }

        v := v1 + v2 + overflow;
        if v >= 10 {
            overflow = 1;
            v = v - 10;
        } else {
            overflow = 0;
        }

        if currNode == nil {
            currNode = &ListNode{Val: v, Next: nil};
            res = currNode;
        } else {
            currNode.Next = &ListNode{Val: v, Next: nil};
            currNode = currNode.Next;
        }



        if node1 != nil {
            node1 = node1.Next
        }
        if node2 != nil {
            node2 = node2.Next
        }
    }

    return res;
}
